File: ai/mcp/python/McpServerBase.py
from contextlib import asynccontextmanager
from collections.abc import AsyncIterator

from pydantic import AnyUrl, TypeAdapter, BaseModel, Field
from typing import Optional, Any, List, Union, Callable, Awaitable, Dict

from mcp.server.fastmcp import FastMCP
from mcp.server.fastmcp.tools.base import Tool
from mcp.server.fastmcp.resources import Resource
from mcp.server.fastmcp.resources.types import FunctionResource
from mcp.server.fastmcp.prompts.base import Prompt, PromptArgument, PromptResult
from mcp.server.fastmcp.resources.templates import ResourceTemplate

from .McpTypes import McpTool, McpPrompt, McpResource, McpToolParameters

# Model context protocol server base.
class McpServerBase:
    """
    Model context protocol server base.
    """
    def __init__(self,
                 name: str,
                 version: str,
                 instructions: str,
                 capabilities: Any,
                 stateless: bool = True):
        """
        Args:
            name:    server name
            version:    server version
            instructions:    server instructions
            capabilities:    server capabilities
            stateless:       is http stateless: true: else false (default is true).

        Example:
            name: "weather",
            version: "1.0.0",
            instructions: "Use this server for....",
            capabilities: {
                resources: {},
                tools: {},
                prompts: {}
            }
        """
        self.open = False
        self.logEvent: Callable[[str, str, str, Any], None] | None = None

        self.name = name
        self.version = version
        self.instructions = instructions
        self.capabilities = capabilities
        self.stateless = stateless

        # Create an MCP server
        self.mcp: FastMCP = FastMCP(name=name, 
                                    instructions=instructions, 
                                    stateless_http=stateless, 
                                    json_response=True)
        self.mcp._mcp_server.version = self.version

    def __repr__(self):
        return f"McpServerBase(name={self.name}, " \
            f"instructions={self.instructions}, " \
            f"stateless={self.stateless}, " \
            f"version={self.version})"

    def onEvent(self, event: Callable[[str, str, str, Any], None]) -> None:
        """
        subscribe to the on event.
        Args:
            event:   the log event handler.
        """
        self.logEvent = event

    def hasStarted(self) -> bool:
        """
        has to MCP server started.

        Return:
            true if started; else false.
        """
        return self.open

    def getMcpServer(self) -> FastMCP:
        """
        get the MCP server, used to register Tools, Resource, Prompts.

        Return:
            the MCP server.
        """
        return self.mcp

    def registerTool(self, 
                     name: str, 
                     callback: Callable[..., Any],
                     description: str | None = None,
                     annotations: Any | None = None) -> bool:
        """
        registers a tool with a config object and callback.

        Args:
            name:  the name of the tool
            callback:   the callback function.
            description:    the tool description
            annotations:    additional tool information

        Return:
            true if tool is registered; else false.

        Example:
            "add",
            lambda a, b: ({
                content: [{ type: "text", text: str(a + b) }]
            }),
            "Add two numbers",
            annotations: {
                title: str | None = None  #A human-readable title for the tool.
            }
        """
        result: bool = False
        try:
            self.mcp.add_tool(callback, name, description, annotations)
            result = True
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "tools", "register tool", e)
            #raise

        return result

    def registerResource(self, 
                     name: str, 
                     uri: AnyUrl,
                     callback: Callable[[], Any],
                     description: str | None = None,
                     mimeType: str | None = None) -> bool:
        """
        registers a resource with a config object and callback.

        Args:
            name:  the name of the resource
            uri:    the URI
            callback:   the callback function, does not take any parameters.
            description:    the resource description
            mimeType:    the mime type

        Return:
            true if resource is registered; else false.

        Example:
            "config",
            "config://app",
            lambda uri: ({
                contents: [{
                    uri: uri.href,
                    text: "App configuration here"
                }]
            }),
            "Application configuration data",
            "text/plain"
        """
        result: bool = False
        try:
            # new resource 
            functionResource: FunctionResource = FunctionResource(
                uri = uri,
                name = name,
                description = description,
                mime_type = mimeType,
                fn = callback
            )

            # create from function.
            self.mcp.add_resource(functionResource)
            result = True
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "resources", "register resource", e)
            #raise

        return result

    def registerResourceTemplate(self, 
                     name: str, 
                     uri: AnyUrl,
                     callback: Callable[..., Any],
                     description: str | None = None,
                     mimeType: str | None = None) -> bool:
        """
        registers a resource template with a config object and callback.

        Args:
            name:  the name of the resource
            uri:    the URI
            callback:   the callback function.
            description:    the resource description
            mimeType:    the mime type

        Return:
            true if resource is registered; else false.

        Example:
            "config",
            "config://app",
            lambda (uri, { userId }) => ({
                contents: [{
                    uri: uri.href,
                    text: f"Profile data for user {userId}"
                }]
            }),
            "Application configuration data",
            "text/plain"
        """
        result: bool = False
        try:
            # new resource 
            resourceTemplate: ResourceTemplate = self.mcp._resource_manager.add_template(
                callback, uri, name, description, mimeType)
            result = True
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "resource_templates", "register resource template", e)
            #raise

        return result

    def registerPrompt(self, 
                     name: str, 
                     callback: Callable[..., PromptResult | Awaitable[PromptResult]],
                     description: str | None = None,
                     argsSchema: List[PromptArgument] | None = None) -> bool:
        """
        registers a prompt with a config object and callback.

        Args:
            name:  the name of the prompt
            callback:   the callback function.
            description:    the prompt description
            argsSchema:    args schema

        Return:
            true if prompt is registered; else false.

        Example:
            "review-code",
            lambda code: ({
                messages: [{
                    role: "user",
                    content: {
                        type: "text",
                        text: f"code: {code}"
                    }
                }]
            }),
            "Review code for best practices and potential issues",
            [
                {
                    name: string;
                    description?: string;
                    required?: boolean; 
                }
            ]
        """
        result: bool = False
        try:
            self.mcp.add_prompt(Prompt(
                name = name,
                description = description,
                arguments = argsSchema,
                fn = callback
            ))
            result = True
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "prompts", "register prompt", e)
            #raise

        return result

    def setToolParameters(self, name: str, parameters: Dict[str, Any] = Field(description="JSON schema for tool parameters")) -> None:
        """
        set the tool parameters

        Args:
            name:  the name of the tool
            parameters: JSON schema for tool parameters

        Example:
        {
            "type": "object",
            "properties": {
                "expression": {
                    "type": "string",
                    "description": "the mathematical expression"
                }
            },
            "required": ["expression"],
            "additionalProperties": False
        }
        """
        tool: Tool = self.mcp._tool_manager.get_tool(name)
        tool.parameters = parameters

    async def getTools(self) -> List[McpTool]:
        """
        get the list of tools

        Return:
            the list of tools; else empty.
        """
        tools: List[McpTool] = [];

        try:
            # load all tools.
            toolsResult = await self.mcp.list_tools()
            if toolsResult is not None:
                for tool in toolsResult:
                    
                    # get base tool parameters.
                    toolParm: Tool = self.mcp._tool_manager.get_tool(tool.name)
                    parameters: Dict[str, Any] | None = None
                    if toolParm.parameters is not None:
                        parameters = toolParm.parameters

                    # create the tool model.
                    tools.append(McpTool(
                        tool.name,
                        tool.name,
                        tool.description,
                        tool.inputSchema,
                        McpToolParameters(parameters)
                    ))
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "tools", "get tools", e)
        
        return tools

    async def getPrompts(self) -> List[McpPrompt]:
        """
        get the list of prompts

        Return:
            the list of prompts; else empty.
        """
        prompts: List[McpPrompt] = [];
        
        try:
            # load all prompts.
            promptsResult = await self.mcp.list_prompts()
            if promptsResult is not None:
                for prompt in promptsResult:
                    # create the prompt model.
                    prompts.append(McpPrompt(
                        prompt.name,
                        prompt.name,
                        prompt.description,
                        prompt.arguments
                    ))
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "prompts", "get prompts", e)

        return prompts

    async def getResources(self) -> List[McpResource]:
        """
        get the list of resources

        Return:
            the list of resources; else empty.
        """
        resources: List[McpResource] = [];

        try:
            # load all resources.
            resourcesResult = await self.mcp.list_resources()
            if resourcesResult is not None:
                for resource in resourcesResult:
                    # create the resource model.
                    resources.append(McpResource(
                        resource.name,
                        resource.name ,
                        resource.description,
                        resource.uri,
                        resource.mimeType
                    ))
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "resources", "get resources", e)

        try:
            # load all resources.
            resourcesResultTemplate = await self.mcp.list_resource_templates()
            if resourcesResultTemplate is not None:
                for resourceTemplate in resourcesResultTemplate:
                    # create the resource model.
                    resources.append(McpResource(
                        resourceTemplate.name,
                        resourceTemplate.name ,
                        resourceTemplate.description,
                        resourceTemplate.uriTemplate,
                        resourceTemplate.mimeType
                    ))
        except Exception as e:
            if (self.logEvent):
                self.logEvent("error", "resource_templates", "get resource templates", e)

        return resources

    async def callTool(self, name: str, args: Dict[str, Any] | None = None) -> Any | None:
        """
        call the tool.

        Args:
            name:    the name of the tool
            args:    the arguments

        Return:
            the result; else none.
        """
        return await self.mcp.call_tool(name, arguments = args)

    async def callPrompt(self, name: str, args: Dict[str, str] | None = None) -> Any | None:
        """
        read the resource.

        Args:
            name:    the name of the prompt
            args:    the arguments

        Return:
            the result; else none.
        """
        return await self.mcp.get_prompt(name, arguments = args)

    async def callResource(self, uri: AnyUrl) -> Any | None:
        """
        read the resource.

        Args:
            uri:    the resource URI

        Return:
            the result; else none.
        """
        return await self.mcp.read_resource(uri)

    def stopServer(self):
        """
        stop receiving messages on stdin and sending messages on stdout
        """
        # if open.
        if self.open:
            self.mcp = None
            self.logEvent = None
            self.open = False

    def startServerStdio(self):
        """
        start receiving messages on stdin and sending messages on stdout.
        For command-line tools and direct integrations.
        """
        # if not open.
        if not self.open:
            try:
                # ... set up server resources, tools, and prompts ...
                # before starting server.
                self.mcp.run(transport='stdio')

                # connection open.
                self.open = True

            except Exception as e:
                self.open = False
                if (self.logEvent):
                    self.logEvent("error", "start", "start server stdio", e)
                #raise  # Re-throws the same exception

    def startServerHttp(self):
        """
        start receiving messages on streamable HTTP.
        For remote servers, set up a Streamable HTTP transport that handles
        both client requests and server-to-client notifications.
        """
        # if not open.
        if not self.open:
            try:
                # ... set up server resources, tools, and prompts ...
                # before starting server.
                self.mcp.run(transport="streamable-http")

                # connection open.
                self.open = True

            except Exception as e:
                self.open = False
                if (self.logEvent):
                    self.logEvent("error", "start", "start server http", e)
                #raise  # Re-throws the same exception

File: ai/mcp/typescript/McpServerBase.ts
import {
    McpServer,
    RegisteredTool,
    RegisteredPrompt,
    RegisteredResource,
    RegisteredResourceTemplate,
    ResourceTemplate
} from "@modelcontextprotocol/sdk/server/mcp.js";
import {
    StdioServerTransport
} from "@modelcontextprotocol/sdk/server/stdio.js";
import {
    StreamableHTTPServerTransport
} from "@modelcontextprotocol/sdk/server/streamableHttp.js";

import { randomUUID } from "node:crypto";

import {
    McpTool,
    McpToolParameters,
    McpPrompt,
    McpResource,
    McpToolCallback,
    McpPromptCallback,
    McpResourceCallback,
    McpHttpTransportModel
} from "./McpTypes.js";

/**
 * Model context protocol server base.
 */
export class McpServerBase {

    // global.
    private open: boolean;
    private mcp: McpServer;
    private httpTransportStateless: boolean;
    private transport: StdioServerTransport | null = null;
    private httpTransports: Array<McpHttpTransportModel>;
    private tools: Array<McpTool>;
    private prompts: Array<McpPrompt>;
    private resources: Array<McpResource>;
    private toolsCallback: Array<McpToolCallback>;
    private promptsCallback: Array<McpPromptCallback>;
    private resourcesCallback: Array<McpResourceCallback>;

    // event logging function.
    private logEvent: (eventType: string, name: string, message: string, details: any) => void;

    /**
     * Model context protocol server base.
     * @param {string} name     server name.
     * @param {string} version  server version.
     * @param {string} instructions  server instructions.
     * @param {object} capabilities  server capabilities.
     * @example
     *      name: "weather",
            version: "1.0.0",
            instructions: "Use this server for....",
            capabilities: {
                resources: {},
                tools: {},
                prompts: {}
            }
     * 
     */
    constructor(public name: string, public version: string, instructions: string,
        public capabilities: {
            resources: {},
            tools: {},
            prompts: {}
        }
    ) {
        // create the server.
        this.open = false;
        this.httpTransportStateless = true;
        this.httpTransports = [];
        this.mcp = new McpServer(
            {
                name: name,
                version: version
            },
            {
                instructions: instructions,
                capabilities: capabilities
            });

        // init
        this.tools = [];
        this.prompts = [];
        this.resources = [];
        this.toolsCallback = [];
        this.promptsCallback = [];
        this.resourcesCallback = [];
        this.logEvent = null;
    }

    /**
     * Subscribe to the on event.
     *
     * @param {function}	event callback(eventType, name, message, details).
     */
    onEvent(event: (eventType: string, name: string, message: string, details: any) => void): void {
        // assign the event.
        this.logEvent = event;
    }

    /**
     * set http transport stateless value: defualt is true (no state).
     * @param {boolean} stateless  set the http transport stateless value; true if stateless; else has state.
     */
    setHttpTransportStateless(stateless: boolean): void {
        this.httpTransportStateless = stateless;
    }

    /**
     * has to MCP server started.
     * @returns {boolean} true if started; else false.
     */
    hasStarted(): boolean {
        return this.open;
    }

    /**
     * get the MCP server, used to register Tools, Resource, Prompts.
     * @returns {McpServer} the MCP server
     */
    getMcpServer(): McpServer {
        return this.mcp;
    }

    /**
     * get the list of mcp http transports.
     * @returns {McpHttpTransportModel} the mcp http transports
     */
    getHttpTransports(): Array<McpHttpTransportModel> {
        return this.httpTransports;
    }

    /**
     * find an MCP Http transport model.
     * @param {string} sessionId   the unique sessionId.
     * @returns {McpHttpTransportModel} the mcp http transport model
     */
    findHttpTransport(sessionId: string): McpHttpTransportModel | null {
        return this.httpTransports.find(function (item) {
            return (item.sessionId === sessionId || item.sessionId === sessionId);
        })
    }

    /**
     * registers a tool with a config object and callback.
     * @param {string} name         the name of the tool
     * @param {object} config       the tool configuration   
     * @param {Function} callback   the callback function.
     * @returns {boolean} true if register; else false.
     * @example 
     *  Add an addition tool:
     *  "add",
        {
            title: "Addition Tool",
            description: "Add two numbers",
            inputSchema: { a: z.number(), b: z.number() }
        },
        async ({ a, b }) => ({
            content: [{ type: "text", text: String(a + b) }]
        })
     */
    registerTool(
        name: string,
        config: {
            title?: string;
            description?: string;
            inputSchema?: any;
            outputSchema?: any;
            annotations?: any;
        },
        callback: any,
        parameters?: McpToolParameters
    ): boolean {
        let result: boolean = false;
        try {
            let registeredTool: RegisteredTool = this.mcp.registerTool(
                name,
                {
                    title: config.title,
                    description: config.description,
                    inputSchema: config.inputSchema,
                    outputSchema: config.outputSchema,
                    annotations: config.annotations
                },
                callback
            );

            // add the tool
            this.tools.push({
                name: name,
                title: config.title,
                description: config.description,
                inputSchema: config.inputSchema,
                parameters: parameters
            });
            this.toolsCallback.push({
                name: name,
                callback: registeredTool.callback
            });
            result = true;
        } catch (e) {
            if (this.logEvent) this.logEvent("error", "tools", "register tool", e);
        }
        return result;
    }

    /**
     * registers a resource with a config object and callback.
     * @param {string} name         the name of the resource
     * @param {string} uri    the URI.
     * @param {object} config   the resource configuration 
     * @param {Function} callback   the callback function.
     * @returns {boolean} true if register; else false.
     * @example 
     *  Application Config:
     *  "config",
     *  "config://app",
        {
            title: "Application Config",
            description: "Application configuration data",
            mimeType: "text/plain"
        },
        async (uri) => ({
            contents: [{
                uri: uri.href,
                text: "App configuration here"
            }]
        })
     */
    registerResource(
        name: string,
        uri: string,
        config: {
            title?: string;
            description?: string;
            mimeType?: string;
        },
        callback: any
    ): boolean {
        let result: boolean = false;
        try {
            let registeredResource: RegisteredResource | RegisteredResourceTemplate = this.mcp.registerResource(
                name,
                uri,
                {
                    title: config.title,
                    description: config.description,
                    mimeType: config.mimeType
                },
                callback
            );

            // add the resource
            this.resources.push({
                name: name,
                title: config.title,
                description: config.description,
                uri: uri,
                mimeType: config.mimeType
            });
            this.resourcesCallback.push({
                name: name,
                callback: registeredResource.readCallback
            });
            result = true;
        } catch (e) {
            if (this.logEvent) this.logEvent("error", "resources", "register resource", e);
        }
        return result;
    }

    /**
     * registers a resource template with a config object and callback.
     * @param {string} name         the name of the resource
     * @param {ResourceTemplate} template    the Template.
     * @param {object} config   the resource configuration 
     * @param {Function} callback   the callback function.
     * @returns {boolean} true if register; else false.
     * @example 
     *  Application Config:
     *  "config",
     *  new ResourceTemplate("users://{userId}/profile", { list: undefined }),,
        {
            title: "User Profile",
            description: "User profile information",
            mimeType: "text/plain"
        },
        async (uri, { userId }) => ({
            contents: [{
                uri: uri.href,
                text: `Profile data for user ${userId}`
            }]
        })
     */
    registerResourceTemplate(
        name: string,
        template: ResourceTemplate,
        config: {
            title?: string;
            description?: string;
            mimeType?: string;
        },
        callback: any
    ): boolean {
        let result: boolean = false;
        try {
            let registeredResource: RegisteredResource | RegisteredResourceTemplate = this.mcp.registerResource(
                name,
                template,
                {
                    title: config.title,
                    description: config.description,
                    mimeType: config.mimeType
                },
                callback
            );

            // add the resource
            this.resources.push({
                name: name,
                title: config.title,
                description: config.description,
                uri: template,
                mimeType: config.mimeType
            });
            this.resourcesCallback.push({
                name: name,
                callback: registeredResource.readCallback
            });
            result = true;
        } catch (e) {
            if (this.logEvent) this.logEvent("error", "resource_templates", "register resource template", e);
        }
        return result;
    }

    /**
     * registers a prompt with a config object and callback.
     * @param {string} name         the name of the prompt
     * @param {object} config   the prompt configuration 
     * @param {Function} callback   the callback function.
     * @returns {boolean} true if register; else false.
     * @example 
     *  Code Review:
     *  "review-code",
        {
            title: "Code Review",
            description: "Review code for best practices and potential issues",
            argsSchema: { code: z.string() }
        },
        ({ code }) => ({
            messages: [{
                role: "user",
                content: {
                    type: "text",
                    text: `Please review this code:\n\n${code}`
                }
            }]
        })
     */
    registerPrompt(
        name: string,
        config: {
            title?: string;
            description?: string;
            argsSchema?: any;
        },
        callback: any
    ): boolean {
        let result: boolean = false;
        try {
            let registeredPrompt: RegisteredPrompt = this.mcp.registerPrompt(
                name,
                {
                    title: config.title,
                    description: config.description,
                    argsSchema: config.argsSchema
                },
                callback
            );

            // add the prompt
            this.prompts.push({
                name: name,
                title: config.title,
                description: config.description,
                arguments: config.argsSchema
            });
            this.promptsCallback.push({
                name: name,
                callback: registeredPrompt.callback
            });
            result = true;
        } catch (e) {
            if (this.logEvent) this.logEvent("error", "prompts", "register prompt", e);
        }
        return result;
    }

    /**
     * get the list of tools
     * @returns {McpTool[]} the list of tools; else empty.
     */
    getTools(): Array<McpTool> {
        return this.tools;
    }

    /**
     * get the list of prompts
     * @returns {McpPrompt[]} the list of prompts; else empty.
     */
    getPrompts(): Array<McpPrompt> {
        return this.prompts;
    }

    /**
     * get the list of resources
     * @returns {McpResource[]} the list of resources; else empty.
     */
    getResources(): Array<McpResource> {
        return this.resources;
    }

    /**
     * call the tool.
     * @param {string} name     the name of the tool
     * @param {object} args     the arguments.
     * @returns {object} the result; else null.
     */
    async callTool(name: string, args: any): Promise<any | null> {

        // try find the tool.
        let mcpToolCallback: McpToolCallback = this.toolsCallback.find(function (item) {
            return (item.name === name || item.name === name);
        })

        // call the tool.
        return await Promise.resolve(mcpToolCallback.callback(args, null));
    }

    /**
     * get the prompt
     * @param {string} name     the name of the prompt
     * @param {object} args     the arguments.
     * @returns {object} the result; else null.
     */
    async callPrompt(name: string, args: any): Promise<any | null> {

        // try find the prompt.
        let mcpPromptCallback: McpPromptCallback = this.promptsCallback.find(function (item) {
            return (item.name === name || item.name === name);
        })

        // call the tool.
        return await Promise.resolve(mcpPromptCallback.callback(args, null));
    }

    /**
     * read the resource.
     * @param {string} name     the name of the resource
     * @param {string} uri  the resource URI.
     * @param {object} args     the arguments if any.
     * @returns {object} the result; else null.
     */
    async callResource(name: string, uri: string, args?: any): Promise<any | null> {

        // try find the resource.
        let mcpResourceCallback: McpResourceCallback = this.resourcesCallback.find(function (item) {
            return (item.name === name || item.name === name);
        })
        let mcpResource: McpResource = this.resources.find(function (item) {
            return (item.name === name || item.name === name);
        })

        // find the uri type
        if (typeof mcpResource.uri === "string") {
            // call the tool.
            return await Promise.resolve(mcpResourceCallback.callback(new URL(uri), null, null));
        }
        else {
            // call the tool.
            return await Promise.resolve(mcpResourceCallback.callback(new URL(uri), args, null));
        }
    }

    /**
     * stop receiving messages.
     */
    async stopServer(): Promise<void> {
        // if open.
        if (this.open) {

            // close all transports
            for (var i = 0; i < this.httpTransports.length; i++) {
                try {
                    // close the transport.
                    if (this.httpTransports[i].transport) await this.httpTransports[i].transport.close();
                } catch (e) {
                    if (this.logEvent) this.logEvent("error", "close", "stop http transport", e);
                }
            }

            try {
                // close the stdio transport.
                if (this.transport) await this.transport.close();

            } catch (e) {
                // some error on close.
                if (this.logEvent) this.logEvent("error", "close", "stop stdio transport", e);
            }
        }

        try {
            // close the mcp connection.
            await this.mcp.close();
        } catch (e) {
            // some error on close.
            if (this.logEvent) this.logEvent("error", "close", "stop server", e);
        }

        // empty list.
        this.httpTransports = [];

        // init
        this.tools = [];
        this.prompts = [];
        this.resources = [];
        this.toolsCallback = [];
        this.promptsCallback = [];
        this.resourcesCallback = [];

        this.open = false;
        this.logEvent = null;
    }

    /**
     * start receiving messages on stdin and sending messages on stdout.
     * For command-line tools and direct integrations.
     */
    async startServerStdio(): Promise<void> {

        // if not open.
        if (!this.open) {
            try {
                // ... set up server resources, tools, and prompts ...
                // before starting server.
                this.transport = new StdioServerTransport();
                await this.mcp.connect(this.transport);
                
                // connection open.
                this.open = true;

            } catch (e) {
                this.open = false;
                if (this.logEvent) this.logEvent("error", "start", "start server stdio", e);
                throw e;
            }
        }
    }

    /**
     * start receiving messages on streamable HTTP.
     * For remote servers, set up a Streamable HTTP transport that handles 
     * both client requests and server-to-client notifications.
     */
    async startServerHttp(): Promise<void> {

        // if not open.
        if (!this.open) {
            try {
                let self = this;

                // ... set up server resources, tools, and prompts ...
                // before starting server.

                // if no state
                if (!this.httpTransportStateless) {

                    // init new transport.
                    const transport: StreamableHTTPServerTransport = new StreamableHTTPServerTransport({
                        sessionIdGenerator: () => randomUUID(),
                        onsessioninitialized: (sessionId) => {
                            // Store the transport by session ID
                            self.httpTransports.push({
                                sessionId: sessionId,
                                transport: transport
                            });
                        }
                    });

                    // Clean up transport when closed
                    transport.onclose = () => {
                        if (transport.sessionId) {
                            try {
                                let httpTransportModel: McpHttpTransportModel = self.findHttpTransport(transport.sessionId);
                                if (httpTransportModel) {

                                    // Get the index of the current transport.
                                    let peerIndex = self.httpTransports.indexOf(httpTransportModel);
                                    if (peerIndex > -1) {
                                        self.httpTransports.splice(peerIndex, 1);
                                    }
                                }
                            }
                            catch (e) {
                                if (this.logEvent) this.logEvent("error", "start", "start server http", e);
                            }
                        }
                    };

                    // Connect to the MCP server
                    await this.mcp.connect(transport);
                }
                else {
                    // stateless
                    // init new transport.
                    const transport: StreamableHTTPServerTransport = new StreamableHTTPServerTransport({
                        sessionIdGenerator: undefined
                    });

                    // add this transport.
                    this.httpTransports.push({
                        sessionId: "sessionid",
                        transport: transport
                    });

                    // Connect to the MCP server
                    await this.mcp.connect(transport);
                }
                
                // connection open.
                this.open = true;

            } catch (e) {
                this.open = false;
                if (this.logEvent) this.logEvent("error", "start", "start server http", e);
                throw e;
            }
        }
    }
}